WIP: Monotonic path handlng

Screwtapello opened the pull request:
Adds a "monotonic_path_preview" module containing my monotonic path handling code, for ease of review, as requested in #17. The module is made public, so the documentation can be viewed with `cargo doc`.

vitiral commented:
sorry, I've been meaning to get to this but work has been busy. I hopefully will shortly.

Screwtapello commented:
It's cool, I've got other stuff to work on too. :)

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+/// or the given path includes a syntactically invalid prefix.
+///
+/// [`Error::IoError`]: struct.Error.html#variant.IoError
+fn split_head_and_tail<P: AsRef<path::Path>>(
```
this seems like it would be generally useful!

Screwtapello replied:
It does seem useful, although I can't think of any concrete use-cases off the top of my head. In this PR, however, I only wanted to demonstrate the `Absolute` and `Relative` structs, though, so I left this as an implementation detail rather than proposing it as a new API.

Feel free to add it to `path_abs` independently, though, if it takes your fancy!

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+
+    debug!("Raw head: {:?}", head);
+
+    // The path's head can be safely converted into an absolute
```
unfortunately this isn't true for windows. See my implementation in `arc::handle_prefix` (which is messy and poor). Windows is pure evil on this.

Here's the rust issue I opened: https://github.com/rust-lang/rust/issues/49342

Screwtapello replied:
I've been using [the official Win10 test VM](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) for VirtualBox—it's supposed to be for testing browsers, but it's a full Win10 install so you can mount your home directory and run binaries and whatever. The only catch is that it only runs for 90 days, but in VirtualBox you can take a system snapshot before the first boot and just revert back to get another 90 days without having to download a fresh image.

Can you point me to an Appveyor build that failed with this problem?

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+        })
+}
+
+/// An absolute path that may or may not exist.
```
Forgive me if I'm mistaken, but I believe a briefer description (since `PathAbs` already exists) is:

> Returns a `PathAbs` by directly following relative directory links while preserving symlinks.

IMO if this were to be included in `path_abs` it would be as a method `PathAbs::follow_relative(path: AsRef<Path>)` (am very open to better names).

Am I accurate in that and does that path forward (lol, pun) make sense?

The point I'm trying to make is that this _is_ a `PathAbs` (it follows the same rules) it is just constructed differently when there are relative paths involved in the constructor path.

Screwtapello replied:
Yes, I expected/hoped that you would take `Absolute::new()`, rename it, and paste it into `PathAbs` as an alternative constructor. Or even (since this is my favourite path-absolutizing algorithm) as the default constructor, `new()`—but I leave that decsion up to you.

I think "relative" is a confusing term, since `foo/bar` is also a relative path and `PathAbs` and `Absolute` treat it the same way. The key difference between them is the treatment of `ParentDir` path components, so I'd probably name the constructor something like `PathAbs::with_parents_resolved()`.

vitiral replied:
Having `PathAbs::with_parents_resolved<P: AsRef<Path>>(path: P)` sounds good to me :+1:

Screwtapello replied:
I think I should start moving code into `src/abs.rs`, using the standard error enum, etc. rather than keeping this in a separate module. Does that sound reasonable?

Screwtapello commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+//!     Relative::new("share/man")?;
+//! ];
+//!
+//! let raw_install_path = std::env::os_args().next()?;
```
Yeah, I guess I meant `std::env::args_os().skip(1).next()`. Like I said, lots of broken doctests here. :(

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+        // If we successfully got metadata, check if it's a symlink.
+        .map(|metadata| metadata.file_type().is_symlink())
+        // If we failed to get metadata...
+        .or_else(|err| {
```
This kind of error handling is always confusing to me. Can this block be broken up into a few statements so it is easier to follow?

Maybe something like:
```
let path = PathArc::from(path);
let meta = path.symlink_metadata()?; // error conversion handled for you
if meta.is_symlink() {
    // ... stuff
} else {
    // other stuff
}
```

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+        // `tail` contains the path components we've yet to check and clean up.
+        let (mut res, mut tail) = split_head_and_tail(path)?;
+
+        let mut seen_paths = collections::BTreeSet::new();
```
currious why a BTreeSet over a conventional HashSet? I guess because the number of items are normally small so it has better cache efficiency?

Screwtapello replied:
Yes. I have no idea if it's actually more efficient, but I had to choose one or the other and "hashing is inefficient for small N" seemed like a reasonable justification. If it turns out to be a bottleneck, it's easy enough to change.

Would you like me to add that justification as a comment?

vitiral replied:
nah, I think it looks reasonable :+1:

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+        let mut seen_paths = collections::BTreeSet::new();
+
+        // While there's still components left to check...
+        while tail.len() > 0 {
```
Please break up the logic in this loop into separate functions. It's very hard to read.

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+                // But first, let's check for symlink loops. If we've already
+                // dereferenced this symlink before, we've hit a loop and we
+                // will never find a suitable answer.
+                if seen_paths.contains(&res) {
```
this can be done in one operation:

```
if let Some(dup) = seen_paths.insert(res.clone()) {
    return Err(... dup);
}
```

Screwtapello replied:
That does mean we always call `res.clone()` even if we don't need to, but I guess it's not a waste in the common case (adding a never-before-seen path to `seen_paths`), so the brevity is worth it.

vitiral commented on src/monotonic_path_preview/mod.rs:
```diff
@@ -0,0 +1,911 @@
+                        // checked all the components from target. Since we're
+                        // pushing to the front, we're pushing components in
+                        // reverse order, and so we have to push this first.
+                        tail.push_front("..".into());
```
this logic is confusing to me. We've got three moving pieces here that we are mutating: `tail`, `res` _and_ `target`. I'm having trouble keeping track of what we are trying to do.

Can this logic be simplified? Maybe a few functions that break it up and don't mutate multiple variables?

Screwtapello replied:
The basic approach is pretty simple: we take an input path that we don't trust, we check each component and move it to the "trusted" pile, and once we've checked everything we return the trusted pile.

For example, let's take a path like "/foo/bar":

1. We split the path into the root ("/") and the tail ("foo", "bar")
1. We assume the root always exists, so it automatically goes into the `res`ult we're going to return.

        res: "/"
        tail: "foo", "bar"

1. We look at the first element in `tail`: "foo" looks reasonable, so we move it across to `res`, and now we have

        res: "/", "foo"
        tail: "bar"

1. We look at the first element in `tail`: "bar" looks reasonable, so we move it across to `res`, and now we have

        res: "/", "foo", "bar"
        tail:

1. `tail` is now empty, so we return `res`

That's a nice, straight-forward loop, easy to understand. However, things get more complex once ".." components enter the picture. Let's say we're given the path "/foo/../bar" where "/foo" is a symlink to "xyz":

1. As before, we split the path, check the components of `tail` and move them across to `res` if they look reasonable, until we get to

        res: "/", "foo"
        tail: "..", "bar"

1. The first component in `tail` is "..", which means the last component of `res` is immediately suspect, so we pop it off and dereference it, getting "xyz", which we shove onto `tail` so it will get checked in due course

        res: "/"
        tail: "xyz", "..", "bar"

1. "xyz" looks reasonable

        res: "/", "xyz"
        tail: "..", "bar"

1. The first component of `tail` is ".." again, so we pop off the last component of `res`, but this time it doesn't dereference to anything interesting

        res: "/"
        tail: "bar"

1. "bar" looks reasonable

        res: "/", "bar"
        tail:

1. `tail` is now empty, so we return `res`

Does that make sense? Things get more complex if a symlink target contains multiple new components, and even more complex if the target is an absolute path instead of a relative one. Rust's type-safety and manual memory management adds complexity on top of that, but the above is the general idea.

I'll think about how I might structure the code to make things more obvious.

Screwtapello commented:
> As a general change, can you remove logging/printing. It doesn't belong in a library this low level.

I usually try to err on the side of "more logging" than "less logging", because it's exactly this kind of "high-level abstraction meets low-level reality" code where mistaken assumptions cause the most confusion. Having `env_logger` set up in tests has definitely helped me understand each test-case, and I expect it would be no less helpful in production.

That said, it's your project, so I'll remove the logging.

vitiral commented:
I _think_ this is one:

https://ci.appveyor.com/project/vitiral/path-abs/build/1.0.88/job/7agq9yrqo2lfw024

```
test_absolute' panicked at 'called `Result::unwrap()` on an `Err` value: Error<The filename, directory name, or volume label syntax is incorrect. (os error 123) when canonicalizing \>'
```

There is also a comment in the existing tests about my confusion:
```
#[cfg_attr(windows, test)]
fn canonicalize_root() {
    // TODO:
    // This works here, but it doesn't work when called in src/arc.rs::handle_prefix
    // during `test_absolute`, possibly related to setting the current_dir?
    expect_path!(r"\\?\C:\", r"\");
}
```

vitiral commented:
Yes, I forgot to mention -- with the changes I suggested I would love to accept this as a "real" PR and get it merged :smile: 

I really like the flexibility this feature will give. Sorry I have so little time, today is my first day at Google!

Screwtapello commented:
> Sorry I have so little time, today is my first day at Google!

Congratulations!

I'll keep poking away, and push a new update when I've addressed your concerns.

Screwtapello commented:
> I *think* this is one:
>
> https://ci.appveyor.com/project/vitiral/path-abs/build/1.0.88/job/7agq9yrqo2lfw024

Thanks for the link!

If I understand the log correctly, it's trying to build commit 3c03806, which [GitHub knows about](https://github.com/vitiral/path_abs/commit/3c0380651573600760729168730cc30b96471bcd) but which apparently isn't included when I `git clone` the repo. Looking at the output of `git log`, however, I think that's the bors equivalent of e6b8b31, which I *did* receive in my `git clone`.

I checked out that revision and built the tests for the target "x86_64-pc-windows-gnu", then booted up my Win10 VM. It turns out the tests expect to be run from somewhere in the C: drive, so once I copied them into the right place, they all passed - even "test_absolute".

It occurred to me that maybe AppVeyor was running a different Windows configuration that was somehow causing problems, so I made [a new repository of my own](https://gitlab.com/Screwtapello/path_abs_test) and uploaded that revision, then signed up for AppVeyor myself, added that repo, and [triggered a build](https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.1)... everything passed. It's quite possible that the AppVeyor image that caused the problem has been replaced in the four months since the problem occurred, so I [tried the VS2013 and VS2017 images as well](https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.2), but everything still succeeded.

AppVeyor apparently publish new VM images every few weeks, and keep the previous images around for a little while in case of regression. The previous images don't date back four months, but I figured I'd [try them all](https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.3) just in case—same result.

I also looked through [the official image changelog](https://www.appveyor.com/updates/) but didn't see any changes to Windows images since March 24th that sounded obviously related.

vitiral commented:
I felt like i was crazy during those bugs, it sounds like maybe I was!

Thanks for the great triage! Nothing would make me happier than to remove
this horrible code. If you wanted to do that in a single commit with a
really really well written commit message i would accept that!

- sent from my Android

On Wed, Aug 1, 2018, 6:53 AM Screwtapello <notifications@github.com> wrote:

> I *think* this is one:
>
>
> https://ci.appveyor.com/project/vitiral/path-abs/build/1.0.88/job/7agq9yrqo2lfw024
>
> Thanks for the link!
>
> If I understand the log correctly, it's trying to build commit 3c03806
> <https://github.com/vitiral/path_abs/commit/3c0380651573600760729168730cc30b96471bcd>,
> which GitHub knows about
> <https://github.com/vitiral/path_abs/commit/3c0380651573600760729168730cc30b96471bcd>
> but which apparently isn't included when I git clone the repo. Looking at
> the output of git log, however, I think that's the bors equivalent of
> e6b8b31
> <https://github.com/vitiral/path_abs/commit/e6b8b31a70f8e9d03bba3350b229fe9cb26feed6>,
> which I *did* receive in my git clone.
>
> I checked out that revision and built the tests for the target
> "x86_64-pc-windows-gnu", then booted up my Win10 VM. It turns out the tests
> expect to be run from somewhere in the C: drive, so once I copied them into
> the right place, they all passed - even "test_absolute".
>
> It occurred to me that maybe AppVeyor was running a different Windows
> configuration that was somehow causing problems, so I made a new
> repository of my own <https://gitlab.com/Screwtapello/path_abs_test> and
> uploaded that revision, then signed up for AppVeyor myself, added that
> repo, and triggered a build
> <https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.1>...
> everything passed. It's quite possible that the AppVeyor image that caused
> the problem has been replaced in the four months since the problem
> occurred, so I tried the VS2013 and VS2017 images as well
> <https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.2>,
> but everything still succeeded.
>
> AppVeyor apparently publish new VM images every few weeks, and keep the
> previous images around for a little while in case of regression. The
> previous images don't date back four months, but I figured I'd try them
> all
> <https://ci.appveyor.com/project/Screwtapello/path-abs-test/build/1.0.3>
> just in case—same result.
>
> I also looked through the official image changelog
> <https://www.appveyor.com/updates/> but didn't see any changes to Windows
> images since March 24th that sounded obviously related.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vitiral/path_abs/pull/20#issuecomment-409582524>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFVCy4xEVQTNdQ2zoq6LKItz84cKlvUOks5uMbLpgaJpZM4UrLBv>
> .
>

vitiral commented:
@Screwtapello is this branch still relevant at all?

Screwtapello commented:
At the very least, this needs to be substantially rewritten due to other PRs I've made. Let's close this for now, and when I have time I can update it and make a new PR.
